lib/nexthop_group.c

```diff
@@ -349,31 +360,34 @@ static bool nexthop_group_parse_nexthop(struct nexthop *nhop,
 
 	nhop->vrf_id = vrf->vrf_id;
 
-	if (addr->sa.sa_family == AF_INET) {
-		nhop->gate.ipv4.s_addr = addr->sin.sin_addr.s_addr;
-		if (intf) {
-			nhop->type = NEXTHOP_TYPE_IPV4_IFINDEX;
-			nhop->ifindex = ifname2ifindex(intf, vrf->vrf_id);
-			if (nhop->ifindex == IFINDEX_INTERNAL)
-				return false;
-		} else
-			nhop->type = NEXTHOP_TYPE_IPV4;
-	} else {
-		memcpy(&nhop->gate.ipv6, &addr->sin6.sin6_addr, 16);
-		if (intf) {
-			nhop->type = NEXTHOP_TYPE_IPV6_IFINDEX;
-			nhop->ifindex = ifname2ifindex(intf, vrf->vrf_id);
-			if (nhop->ifindex == IFINDEX_INTERNAL)
-				return false;
-		} else
-			nhop->type = NEXTHOP_TYPE_IPV6;
+	if (intf) {
+		nhop->ifindex = ifname2ifindex(intf, vrf->vrf_id);
+		if (nhop->ifindex == IFINDEX_INTERNAL)
+			return false;
 	}
 
+	if (addr) {
+		if (addr->sa.sa_family == AF_INET) {
+			nhop->gate.ipv4.s_addr = addr->sin.sin_addr.s_addr;
+			if (intf)
+				nhop->type = NEXTHOP_TYPE_IPV4_IFINDEX;
+			else
+				nhop->type = NEXTHOP_TYPE_IPV4;
+		} else {
+			nhop->gate.ipv6 = addr->sin6.sin6_addr;
+			if (intf)
+				nhop->type = NEXTHOP_TYPE_IPV6_IFINDEX;
+			else
+				nhop->type = NEXTHOP_TYPE_IPV6;
+		}
+	} else
+		nhop->type = NEXTHOP_TYPE_IFINDEX;
+
 	return true;
 }
 
 DEFPY(ecmp_nexthops, ecmp_nexthops_cmd,
-      "[no] nexthop <A.B.C.D|X:X::X:X>$addr [INTERFACE]$intf [nexthop-vrf NAME$name]",
+      "[no] nexthop {<A.B.C.D|X:X::X:X>$addr|INTERFACE$intf} [nexthop-vrf NAME$name]",
```

Comment:
This is wrong imo

robot(config-nh-group)# nexthop 192.168.11.44 
  <cr>         
  A.B.C.D      v4 Address
  INTERFACE    Interface to use
     docker0 dummy1 enp3s0 lo mveth1 mveth2 mveth3 mveth4 mveth5 mveth6 mveth7 mveth8 mveth9 
     mveth10 mveth11 mveth12 mveth13 mveth14 pimreg tun0 virbr0 virbr0-nic 
  X:X::X:X     v6 Address
  nexthop-vrf  If the nexthop is in a different vrf tell us
robot(config-nh-group)# nexthop 192.168.11.44 
  <cr>         
  A.B.C.D      v4 Address
  INTERFACE    Interface to use
     docker0 dummy1 enp3s0 lo mveth1 mveth2 mveth3 mveth4 mveth5 mveth6 mveth7 mveth8 mveth9 
     mveth10 mveth11 mveth12 mveth13 mveth14 pimreg tun0 virbr0 virbr0-nic 
  X:X::X:X     v6 Address
  nexthop-vrf  If the nexthop is in a different vrf tell us
robot(config-nh-group)# nexthop 192.168.11.44 mveth1 
  <cr>         
  X:X::X:X     v6 Address
  nexthop-vrf  If the nexthop is in a different vrf tell us

We should not have these options presented like this.

Reply:
I agree. This is the CLI bug I reported in #3379, I forgot we still can't use angle brackets nested inside curly braces. I'll try to work around this problem somehow.